x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/jdbc/JdbcPreparedStatement.java

```diff
@@ -271,12 +318,101 @@ public void setNClob(int parameterIndex, Reader reader, long length) throws SQLE
+                || x instanceof Float
+                || x instanceof Double
+                || x instanceof String) 
+        {
+            setParam(parameterIndex, 
+                    TypeConverter.convert(x, TypeConverter.fromJavaToJDBC(x.getClass()), DataType.fromJdbcTypeToJava(targetJDBCType)), 
+                    targetSqlType);
+        } else if (x instanceof Timestamp
+                || x instanceof Calendar
+                || x instanceof Date
+                || x instanceof LocalDateTime
+                || x instanceof Time
+                || x instanceof java.util.Date) {
+            if (targetJDBCType == JDBCType.TIMESTAMP ) {
+                // converting to {@code java.util.Date} because this is the type supported by {@code XContentBuilder} for serialization
+                java.util.Date dateToSet;
+                if (x instanceof Timestamp) {
+                    dateToSet = new java.util.Date(((Timestamp) x).getTime());
+                } else if (x instanceof Calendar) {
+                    dateToSet = ((Calendar) x).getTime();
+                } else if (x instanceof Date) {
+                    dateToSet = new java.util.Date(((Date) x).getTime());
+                } else if (x instanceof LocalDateTime){
+                    LocalDateTime ldt = (LocalDateTime) x;
+                    Calendar cal = getDefaultCalendar();
+                    cal.set(ldt.getYear(), ldt.getMonthValue() - 1, ldt.getDayOfMonth(), ldt.getHour(), ldt.getMinute(), ldt.getSecond());
+                    
+                    dateToSet = cal.getTime();
+                } else if (x instanceof Time) {
+                    dateToSet = new java.util.Date(((Time) x).getTime());
+                } else {
+                    dateToSet = (java.util.Date) x;
+                }
+
+                setParam(parameterIndex, dateToSet, Types.TIMESTAMP);
+            } else if (targetJDBCType == JDBCType.VARCHAR) {
+                setParam(parameterIndex, String.valueOf(x), Types.VARCHAR);
+            } else {
+                // anything else other than VARCHAR and TIMESTAMP is not supported in this JDBC driver
+                throw new SQLFeatureNotSupportedException(
+                        "Conversion from type " + x.getClass().getName() + " to " + targetJDBCType + " not supported");
+            }
+        } else if (x instanceof byte[]) {
+            if (targetJDBCType != JDBCType.VARBINARY) {
+                throw new SQLFeatureNotSupportedException(
+                        "Conversion from type byte[] to " + targetJDBCType + " not supported");
+            }
+            setParam(parameterIndex, x, Types.VARBINARY);
+        } else {
+            throw new SQLFeatureNotSupportedException(
+                    "Conversion from type " + x.getClass().getName() + " to " + targetJDBCType + " not supported");
+        }
+    }
+
+    private void checkKnownUnsupportedTypes(Object x) throws SQLFeatureNotSupportedException {
```

Comment:
I feel like it'd be cleaner to have this be the result of falling off the end of an if statement chain.